src/merkle/mmr/partial.rs

```diff
@@ -0,0 +1,362 @@
+        // the PartialMmr.
+        let computed = path.compute_root(index as u64, node).map_err(MmrError::MerkleError)?;
+        if self.peaks[target_tree] != computed {
+            return Err(MmrError::InvalidPeak);
+        }
+
+        // Compute the leaf position in the PartialMmr, the position is equal to the sum of the
+        // leaves on the bigger trees plus the position of the leaf in its owning tree
+        let mmr_index: usize = (target_forest ^ tree) + index;
+        let mut idx = InOrderIndex::leaf_to_index(mmr_index);
+
+        for node in path.nodes() {
+            self.nodes.insert(idx.sibling(), *node);
+            idx = idx.up();
+        }
+
+        Ok(())
+    }
+
+    /// Remove a leaf of the [ParialMmr] and the unused nodes from the authentication path.
+    ///
+    /// Note: `leaf_pos` corresponds to the position the [Mmr] and not on an individual tree.
+    pub fn remove(&mut self, leaf_pos: usize) {
+        let mut idx = InOrderIndex::leaf_to_index(leaf_pos);
+
+        self.nodes.remove(&idx.sibling());
+
+        // `idx` represent the element that can be computed by the authentication path, because
+        // these elements can be computed they are not saved for the authentication of the current
+        // target. In other words, if the idx is present it was added for the authentication of
+        // another element, and no more elements should be removed otherwise it would remove that
+        // element's authentication data.
+        while !self.nodes.contains_key(&idx) {
+            idx = idx.up();
+            self.nodes.remove(&idx.sibling());
+        }
+    }
+
+    /// Applies updates to the [PartialMmr].
+    pub fn update<T: AsRef<[RpoDigest]>>(
+        &mut self,
+        forest: usize,
+        updates: T,
+    ) -> Result<(), MmrError> {
+        if forest < self.forest {
+            return Err(MmrError::InvalidPeaks);
+        }
+
+        let updates = updates.as_ref();
+        if forest == self.forest && !updates.is_empty() {
+            return Err(MmrError::InvalidUpdate);
+        }
+
+        // find the tree merges
+        let changes = self.forest ^ forest;
+        let largest = 1 << changes.ilog2();
+        let merges = self.forest & (largest - 1);
+
+        debug_assert!(
+            !self.track_latest || (merges & 1) == 1,
+            "if there is an odd element, a merge is required"
+        );
+
+        // keeps track of how many data elements from the update have been consumed
+        let mut update_count = 0;
+
+        if merges != 0 {
+            // starts at the smallest peak and follows the merged peaks
+            let mut peak_idx = forest_to_root_index(self.forest);
+
+            // match order of the update data while applying it
+            self.peaks.reverse();
+
+            // set to true when the data is needed for authentication paths updates
+            let mut track = self.track_latest;
+            self.track_latest = false;
```

Comment:
IMO that is the best approach. I'm assuming most clients wont be following the server and fetching every single update, so they are likely to learn about new notes for a collection of blocks, and having to filter out the notes that come into the odd block would make the calling code more complicated.

OTOH, if we do think there is a use case that will require following the rollup's tip closely, then we can find a middle ground by adding another method that does both update and add? That way I think both use cases would have a nice API to work with.

Reply:
Thinking more about this, it is probably fine to leave as is.

I'm imagining the node will send a sync response which will include a tuple `(mmr_delta, path_to_last_header)`. This `path_to_last_header` would be empty if there were notes of interest in the last returned block header. So, our update code on the client side would look something like this:
```Rust
let latest_block_has_notes = /* computed somewhere */

chain_mmr.update(mmr_delta).unwrap();
if latest_block_has_notes {
    chain_mmr.add(block_num, block_header_hash, path_to_last_header).unwrap();
}
```
And then everything will work fine.

One thing I realized is that we should add another constructor for `PartialMmr` to create it from individual components (not just from peaks).

Reply:
I assumed it would be `(MmrDelta, Vec<(BlockHeader, MerklePath)>)`, because a client that is offline for a while may have multiple notes of interest created, and they are not necessarily from the latest block.

Or perhaps I'm missing something? I assumed that we would receive the `BlockHeader` in which a note was created. Are you saying that we don't have to use the note creation block, but can use any block after the note was created?

Reply:
The full response would look something like this (based on [this](https://github.com/0xPolygonMiden/miden-node/issues/43#issuecomment-1765090819)):
```
Response {
    chain_tip: <number of the latest block in the chain>,
    block_header: <block header of the block with the first note matching the specified criteria>,
    mmr_delta: <data needed to update the partial MMR from `block_ref` to `block_header.block_num`>,
    block_path: <Merkle path to the in chain MMR to the block defined by `block_header`>,
    notes: [a list of all notes together with the Merkle paths from `block_header.note_root`],
}
```
So, the `block_header` is already there, and we also have `notes` array to figure out if the block has notes.

Because of how this endpoint works, we will always get data for a single block that has notes - except for when we already at the tip (`block_header.block_num == chain_tip`). In the latter case, we may or may not get a block with notes.